# GT-Seq Pipeline

## Needed Scripts
We will use the GT-Seq pipeline developed and described by Campbell et al. (2015), with necessary scripts located at [https://github.com/GTseq/GTseq-Pipeline](https://github.com/GTseq/GTseq-Pipeline). Specifically, the scripts that we need are all written in PERL, and are:  

1. `GTseq_HashSeqs.pl`: This script collapses unique reads into a single fasta file and counts the occurrences of each. The resulting `.hash` file cannot be viewed, but is used by the `GTseq_SeqTest.pl` script.   
2. `GTseq_SeqTest.pl`: Script to identify how many times each forward primer and probe sequences occur in each sample.   
3. `GTseq_Genotyper_v3.pl`: Identifies the genotypes of each individual at each locus.    
4. `GTseq_GenoCompile_v3.pl`: Compiles individual genotpes generated by the `Genotype` script.    
5. `GTseq_ErrorReport_v3.pl`: This script summarizes possible null alleles and their sequences within the genotyped samples.    

I have put these in a `scripts` directory within the `Right_Whale/GT-Seq/` directory, so that they can stay there and be referred to for each analysis.


## Needed Files
There are a few other files that we will need for these analyses as well. Ones that have already been created - and which will be the same each time - are:  

1. **NARW_276_Seqtest.txt**, which contains the locus names and primer sequences for each locus (SNPs plus sexing), and  
2.  **NARW_276gtseq_genotyper_input.csv**, which contains much of the same information, but also including the reference and alternative alleles for each position.  

I have put these in a `reference` directory within the `Right_Whale/GT-Seq/` directory, so that they can stay there and be referred to for each analysis.


## Housekeeping
To clean things up a bit, I created a new directory, called `raw` and moved all of the sample files into that, so that the current directory contains a `raw` directory, a `filtered` directory with the files we generated in the previous tutorial, and our **multiqc** results. I did this with the following commands.
```
mkdir raw

mv Sample_* raw/
```


I then created a directory to hold the GT-Seq results.
```
mkdir GT-Seq
```

The GT-Seq scripts only work on `fastq` files, and not on zipped `fastq.gz` files. Therefore, we need to unzip them.
```
gunzip *_merged.fastq.gz
```

Lastly, we need a list of sample names that we can use for looping. We can create this from our already existing `files` file.
```
sed "s/Sample_//g" files > samples
```


Ee will need this `samples` file in the **GT-Seq** directory, so let's copy it to there, and then move into that directory.
```
cp samples ../GT-Seq/samples

cd ../GT-Seq
ls
```


## Pre-Run Check
You may want to do a sanity check prior to running all the scripts on all the files. By this, I mean just checking to see how many times a few primers and/or probe regions exist in some of your files. The scripts below will tell you this, but they don't always behave how you might expect. Therefore, I find it helpful to get a feel for some of these values manually, so that I know what to expect from the scripts, and can compare results across the two different methods.

For example, to count how many times the first forward primer occurs in the first sample, you can run the following command (be copying and pasting the sequences and file names).
```
grep -c "CCTGAAGAGGCTAAGACAGAATGT" ../filtered/EGL00003-2_FRA2523A5-1_merged.fastq
```

The result was **19**. We can also count how many times the probe region occurs (with either SNP), using the following command. Note the use of the pipe `|` to say "OR", and that it first needs to be escaped using the forwward slash (`\`).
```
grep -c "ATCAGTGTAGGGAAT\|ATCAGTGCAGGGAAT" ../filtered/EGL00003-2_FRA2523A5-1_merged.fastq
```

The result was **18**, which is a little surprising. This means that there was one read in this sample that contained the forward primer sequence, but not the "probe" region for the variable site. Regardless, we now know that we expect the first individual to have 18 reads for the first locus.



## Running the Pipeline
Although we could write a single script that would do most of these things one after the other - and that would be faster - I prefer to run them (mostly) one at a time, and check the resulting files in between the steps (when possible), just to make sure that things have gone as expected.

First we will use the `GTseq_HashSeqs.pl` script to generate the `.hash` files containing the counts of each unique sequence in each sample. We cannot view the resulting file, so there is nothing to check after this step. So, in this case, we will also pair this command with the next one, and get these two done at the same time. This will take a while (maybe 5-10 minutes).
```
while read SAMPLE;
do
	echo ${SAMPLE}_merged.fastq;
	perl ../../../scripts/GTseq_HashSeqs.pl ../filtered/${SAMPLE}_merged.fastq > ${SAMPLE}.hash;
	perl ../../../scripts/GTseq_SeqTest.pl ../../../reference/NARW_276_Seqtest.txt ${SAMPLE}.hash > ${SAMPLE}.seqtest.csv;
done < samples
```

The resulting `seqtest.csv` files will contain how many times each locus, and each allele, were found in each individual. We can use the `head` command to look through a few of these and see if they look how we expected. For example:
```
head EGL00003-2_FRA2523A5-1.seqtest.csv
```


Next, we will create run the Genotyper script using a similar loop.

```
while read SAMPLE;
do
	echo ${SAMPLE};
  perl ../../../scripts/GTseq_Genotyper_v3.pl ../../../reference/NARW_276gtseq_genotyper_input.csv ../filtered/${SAMPLE}_merged.fastq > ${SAMPLE}.genos;
done < samples
```

The first row of each resulting `.genos` file contains summary information about the number of reads scored for each locus. It can be helpful to take a look at these before moving forward. This can be done with the following command.
```
head -n1 *.genos
```

## References
Campbell NR, Harmon SA, Narum SH (2015) Genotyping-in-Thousands by sequencing (GT-seq): A cost effective SNP genotyping method based on custom amplicon sequencing. *Molecular Ecology Resources* **15**: 855-867.
