# GT-Seq Pipeline

## Needed Scripts
We will use the GT-Seq pipeline developed and described by Campbell et al. (2015), with necessary scripts located at [https://github.com/GTseq/GTseq-Pipeline](https://github.com/GTseq/GTseq-Pipeline). Specifically, the scripts that we need are all written in PERL, and are:  

1. GTseq_HashSeqs.pl  
2. GTseq_SeqTest.pl  
3. GTseq_Genotyper_v3.pl  
4. GTseq_GenoCompile_v3.pl  
5. GTseq_ErrorReport_v3.pl  

I have put these in a `scripts` directory within the `Right_Whale/GT-Seq/` directory, so that they can stay there and be referred to for each analysis.


## Needed Files
There are a few other files that we will need for these analyses as well. Ones that have already been created - and which will be the same each time - are:  

1. **NARW_276_Seqtest.txt**, which contains the locus names and primer sequences for each locus (SNPs plus sexing), and  
2.  **NARW_276gtseq_genotyper_input.csv**, which contains much of the same information, but also including the reference and alternative alleles for each position.  

I have put these in a `reference` directory within the `Right_Whale/GT-Seq/` directory, so that they can stay there and be referred to for each analysis.


## Housekeeping
To clean things up a bit, I created a new directory, called `raw` and moved all of the sample files into that, so that the current directory contains a `raw` directory, a `filtered` directory with the files we generated in the previous tutorial, and our **multiqc** results. I did this with the following commands.
```
mkdir raw

mv Sample_* raw/
```


I then created a directory to hold the GT-Seq results, and ran all the following commands from this directory.
```
mkdir GT-Seq

cd GT-Seq
```


## One More Needed File
We also need to generate a file that contains a list of the samples that will be analyzed. This should be based on the files that we generated in the previous section (*i.e.,* some files should end in `.merged.fastq.gz`), and the should be located in our `filtered` directory.
```
ls ../filtered/*merged.fastq.gz > samples
```

Check the file.
```
head samples
```


Then we need to remove the `../filtered/` prefix from each sample.
```
sed -i "s,../filtered/,,g" samples

head samples
```

Then we need to remove the `_merged.fastq.gz` suffix from each sample name.
```
sed -i "s/_merged.fastq.gz//g" samples

head samples

wc -l samples
```


## Running the Pipeline
Next, we will create a script that will run the major components of the GT-Seq pipeline. We can call this script `gtseq`.
```
nano gtseq
```

It should contain the following commands.
```
#!/bin/bash

while read SAMPLE;
do
  perl ../../../scripts/GTseq_HashSeqs.pl ../filtered/${SAMPLE}_merged.fastq.gz > ${SAMPLE}.hash;
  
  perl ../../../scripts/GTseq_SeqTest.pl ../../../reference/NARW_276_Seqtest.txt ${SAMPLE}.hash > ${SAMPLE}.seqtest.csv;
  
  perl ../../../scripts/GTseq_Genotyper_v3.pl ../../../reference/NARW_276gtseq_genotyper_input.csv ../filtered/${SAMPLE}_merged.fastq.gz > ${SAMPLE}.genos;
  
done < samples
```

Then make this an executable file.
```
chmod a+x gtseq
```

Then run it.
```
./gtseq
```


## References
Campbell NR, Harmon SA, Narum SH (2015) Genotyping-in-Thousands by sequencing (GT-seq): A cost effective SNP genotyping method based on custom amplicon sequencing. *Molecular Ecology Resources* **15**: 855-867.