# Analysis of mtDNA from GT-Seq

## Needed Files
For these analyses, we need a reference version of the targeted region of the mitochondrial control region that contains ambiguous codes at variable sites. This file is called `mtdna_consensus.fasta` and is stored in the `reference` directory within the `Right_Whale/GT-Seq/` directory, so that it can stay there and be referred to for each analysis.

I will also make a new directory, called **mtDNA** into which we will write results, and then give everyone permission to write to that directory.
```
mkdir mtDNA
chmod -R 777 mtDNA
```


## Mapping Reads
We will use **bwa** to map the reads to the mitochondrial control region, and then use **samtools** to clean things up a bit. The commands below are run from *within* the `filtered` directory, but are written into the `mtDNA` directory.

First, index the reference sequence (don't need to do this every time, but also doesn't hurt).
```
bwa index ../../../reference/mtdna_consensus.fasta
```

Then map the reads (this will take a little while).
```
while read SAMPLE;
do
bwa mem -M -t 1 ../../../reference/mtdna_consensus.fasta ${SAMPLE}_merged.fastq > ../mtDNA/${SAMPLE}_mtdna_gtseq.sam 2> ../mtDNA/${SAMPLE}-mtdna_gtseq.err
done < samples
``` 

Then, go into the mtDNA directory and view one or two of the `.sam` and `.err` files (using `head`) to ensure that everything worked properly.
```
cd ../mtDNA/
```

Then, copy the `samples` list into the `mtDNA` directory.
```
cp ../filtered/samples samples
```

Then use **samtools** or organize the data and write consensus sequences for each individual.
```
while read SAMPLE;
do
samtools sort -o ${SAMPLE}_mtdna_gtseq_sort.bam ${SAMPLE}_mtdna_gtseq.sam;
samtools consensus --show-ins --ambig --min-MQ 30 --min-BQ 30 -o ${SAMPLE}_mtDNA_consensus.fasta ${SAMPLE}_mtdna_gtseq_sort.bam;
samtools coverage ${SAMPLE}_mtdna_gtseq_sort.bam > ${SAMPLE}_coverage.txt;
done < samples
```

Use the `head` commands to view some of the `consensus.fasta` files to make sure that things worked as expected.

The above commands call the consensus for each sample ``Consensus Hap''. We want to replace that with the actual sample name.
```
while read SAMPLE;
do
sed -i "s/ConsensusHap/${SAMPLE}/g" ${SAMPLE}_mtDNA_consensus.fasta;
sed -i "s/ConsensusHap/${SAMPLE}/g" ${SAMPLE}_coverage.txt;
done < samples
```

Use the `head` commands to view some of the `consensus.fasta` files to make sure that things worked as expected.

Lastly, combine all of the consensus fasta files into one larger file that can be viewed in a sequence viewer/alignment tool of your choice.
```
cat *_mtDNA_consensus.fasta > NARW_mtDNA_haplotypes.fasta
```
