# Checking mtDNA Data

## Rationale
There may be times when the mitochondrial sequences differ from what is expected and/or from results that we have obtained before. In these cases, it can be handy to review the raw reads for the samples in question, to see where any issues may have arisen. The commands below allow you to do this.


## Checking Reads for a Sample
What we can do is a 'quick and dirty' look at an alignment of the reads for a sample against the reference sequence. To do this, we need to start with an indexed `.bam` file.  We already have sorted `.bam` files for each sample (generated in earlier steps), but now we need to index them. We can do so with the following command
```
samtools index filename.bam
```

Next, we can compare these reads with the reference file, and show them in a viewable alignment format. This can be done with the `tview` command from **samtools**, where we provide the `.bam` file to use, as well as the reference `.fasta` file for comparison.
```
samtools tview filename.bam ../../../reference/mtdna_consensus.fasta
```

